File: src/main/java/com/leet/medium/WordBreak.ts
function solve(
  str: string,
  words: string[],
  pos: number,
  failed: Set<number>
): boolean {
  if (pos >= str.length) return true;

  if (failed.has(pos)) {
    return false;
  }

  let ans = false;

  for (let word of words) {
    const sub = str.substring(pos, pos + word.length);
    if (sub === word) {
      ans = solve(str, words, pos + word.length, failed);
    }
    if (ans) return ans;
  }

  failed.add(pos);

  return false;
}

function wordBreak(s: string, words: string[]): boolean {
  return solve(s, words, 0, new Set());
}

File: src/main/java/com/leet/medium/WordBreak.go
package main

// 0ms, 100%

var dp []int

type TrieNode struct {
  nodes [26]*TrieNode
  val string
  leaf bool
}

var root *TrieNode

func (trieNode *TrieNode) addWord(word string) {
  node := trieNode
  for _, char := range word {
    if node.nodes[char-'a'] == nil {
      node.nodes[char-'a'] = &TrieNode{}
    }
    node = node.nodes[char-'a']
  }
  node.leaf = true
  node.val = word
}

func wordBreakUtil(s string, pos int) bool {
  if pos >= len(s) {
    return true
  }
  if dp[pos] != 0 {
    return dp[pos] == 1
  }
  node := root
  ans := false
  for i := pos; i < len(s); i++ {
    char := s[i]
    if node.nodes[char-'a'] == nil {
      break
    }
    node = node.nodes[char-'a']
    if node.leaf {
      ans = wordBreakUtil(s, i+1)
    }
    if ans {
      break
    }
  }
  if ans {
    dp[pos] = 1
  } else {
    dp[pos] = -1
  }
  return ans
}

func wordBreak(s string, wordDict []string) bool {
  dp = make([]int, len(s))
  root = &TrieNode{}
  for _, word := range wordDict {
    root.addWord(word)
  }
  return wordBreakUtil(s, 0)
}
